clamav.py

```diff
@@ -195,21 +197,48 @@ def scan_file(path):
         stdout=subprocess.PIPE,
         env=av_env,
     )
-    output = av_proc.communicate()[0].decode()
-    print("clamdscan output:\n%s" % output)
 
-    # Turn the output into a data source we can read
-    summary = scan_output_to_json(output)
+    try:
+        output, errors = av_proc.communicate(timeout=CLAMDSCAN_TIMEOUT)
+    except subprocess.TimeoutExpired:
+        av_proc.kill()
+        output, errors = av_proc.communicate()
+
+    decoded_output = output.decode()
+    print("clamdscan output:\n%s" % decoded_output)
+
     if av_proc.returncode == 0:
         return AV_STATUS_CLEAN, AV_SIGNATURE_OK
     elif av_proc.returncode == 1:
+        # Turn the output into a data source we can read
+        summary = scan_output_to_json(decoded_output)
         signature = summary.get(path, AV_SIGNATURE_UNKNOWN)
         return AV_STATUS_INFECTED, signature
     else:
         msg = "Unexpected exit code from clamdscan: %s.\n" % av_proc.returncode
+
+        if errors:
+            msg += "Errors: %s\n" % errors.decode()
+
         print(msg)
         raise Exception(msg)
 
+def is_clamd_running():
+    clamd_socket = '/tmp/clamd.sock'
+    print("Checking if clamd is running on %s" %clamd_socket)
+
+    if os.path.exists(clamd_socket):
+        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as s:
+            s.connect(clamd_socket)
+            s.send(b'PING')
+            data = s.recv(32)
```

Comment:
My understanding is that sockets are in blocking mode by default. So, it's probably very unlikely, but if the `clamd_socket` doesn't send 32 bytes for whatever reason, we won't fill our buffer, so will continue to wait for output indefinitely.
It might be worth using `settimeout` (https://docs.python.org/3/library/socket.html#socket.socket.settimeout) before calling `connect` (see https://docs.python.org/3/library/socket.html#timeouts-and-the-connect-method), or using `setdefaulttimeout` (https://docs.python.org/3/library/socket.html#socket.setdefaulttimeout).

If we're adding a timeout, we probably want to wrap this in a try/except as well.

```
try:
  data = s.recv(32)
except (socket.timeout, socket.error), e:
  # could retry the communication, but probably best to just return False to denote that clamd isn't running, or we had a problem communicating with it
```

We don't strictly need to handle a timeout or generic error in the same except, but a decent starting point is probably just logging the issue and returning false from this method.

Also, just as a note, I'm seeing from the discussion on this SO (https://stackoverflow.com/questions/2719017/how-to-set-timeout-on-pythons-socket-recv-method), that some people still have trouble with `recv()` hanging even with a timeout, so another possible approach is

```
import select

s.setblocking(0)

ready = select.select([s], [], [], timeout_in_seconds)
if ready[0]:
    data = s.recv(32)
```

However, there seems to be some disagreement on whether this is actually necessary, so I'd recommend we go with setting a timeout and catching the error, and only look into this second approach if we have issues with `recv()` hanging.